js/src/components/shipping-rate-section/minimum-order-card/getMinimumOrderHandlers.js

```diff
@@ -0,0 +1,121 @@
+/**
+ * @typedef { import(".~/data/actions").ShippingRate } ShippingRate
+ * @typedef { import(".~/data/actions").CountryCode } CountryCode
+ * @typedef { import("./typedefs").MinimumOrderGroup } MinimumOrderGroup
+ */
+
+/**
+ * @typedef {Object} getMinimumOrderHandlersResult
+ * @property {(newGroup: MinimumOrderGroup) => void} handleAddSubmit Handler for add new minimum order group.
+ * @property {(oldGroup: MinimumOrderGroup) => (newGroup: MinimumOrderGroup) => void} getChangeHandler Get handler for change minimum order group.
+ * @property {(newGroup: MinimumOrderGroup) => () => void} getDeleteHandler Get handler for delete minimum order group.
+ */
+
+/**
+ * Get handlers for shipping rate minimum order.
+ *
+ * @param {Object} props
+ * @param {Array<ShippingRate>} props.value Shipping rates.
+ * @param {(newValue: Array<ShippingRate>) => void} props.onChange Callback called with new data when minimum order for shipping rates are changed.
+ * @return {getMinimumOrderHandlersResult} Handlers.
+ */
+const getMinimumOrderHandlers = ( { value, onChange } ) => {
```

Comment:
Similarly to https://github.com/woocommerce/google-listings-and-ads/pull/1424#discussion_r860270678:

It's somewhat alerting to me that `getHandlers` returns one `handler` and two `get*Handlers`. `getHandler` that returns `get*Handler` isn't it `getGetHandler` ;) - that sounds quite redundant

Also, what is the benefit of having one function (`getHandlers`) that returns a set of three disjointed functions? They do not need access to each other or share any logic, do they?

Maybe we can just `export { getAddHandler, getChangeHandler, getDeleteHandler }` directly? Then we will save on constructing & destructing, and avoid the naming problem mentioned above. That will also make them more atomic and easier to test and maintain independently.

Reply:
:nail_care: Plus, I'd make it at least 
```suggestion
const getMinimumOrderHandlers = ( { value, onChange } ) => {
```

As we call it as `getMinimumOrderHandlers( { value, onChange } );` anyway, so constructing and deconstructing only blurs the picture for me

Reply:
> It's somewhat alerting to me that `getHandlers` returns one `handler` and two `get*Handlers`. `getHandler` that returns `get*Handler` isn't it `getGetHandler` ;) - that sounds quite redundant

Ahh right, I'll think of a better naming. If you have a suggestion, do share. 😄 

> Also, what is the benefit of having one function (`getHandlers`) that returns a set of three disjointed functions? They do not need access to each other or share any logic, do they?

They don't need access to each other. I was thinking that the three functions are related ("add", "edit" and "delete" operations; part of CRUD operations), and because they all take in `value` and `onChange`, I thought it make sense to group them together. The benefit is that on the caller side, they just need to have one call to `getMinimumOrderHandlers` to get all three of them (instead of three separate calls).

To be more specific, I mean: 

```js
// current way - one call to get all three:
const { getA, getB, getC } = callGetABC( { value, onChange } );

// alternative way 1:
const getA = callGetA( { value, onChange } );
const getB = callGetB( { value, onChange } );
const getC = callGetB( { value, onChange } );

// alternative way 2:
const [ getA, getB, getC ] = [
	callGetA( { value, onChange } ),
	callGetB( { value, onChange } ),
	callGetB( { value, onChange } ),
];
```

> Maybe we can just `export { getAddHandler, getChangeHandler, getDeleteHandler }` directly? Then we will save on constructing & destructing, and avoid the naming problem mentioned above. That will also make them more atomic and easier to test and maintain independently.

That is a good idea! I'll look into it. 😄 👍 

> 💅 Plus, I'd make it at least [...]

There is no change in your suggested change...? 😅 I'm guessing you mean changing it to `getMinimumOrderHandlers ( value, onChange )` (accept two parameters instead of one object parameter)? I was thinking because they were both coming from the react props, it makes sense to keep them in object props structure. WDYT? Open to changes here.

Reply:
> There is no change in your suggested change...? sweat_smile I'm guessing you mean changing it to `getMinimumOrderHandlers ( value, onChange )` (accept two parameters instead of one object parameter)? 

Yes, I fixed the comment above.

>  I was thinking because they were both coming from the react props, it makes sense to keep them in object props structure. WDYT? Open to changes here.

I think in most of the cases data come from React props, and I still don't see an actual benefit in https://github.com/woocommerce/google-listings-and-ads/pull/1410/files#diff-19741a9dbe0895f99f4b04a6a88e21d3a8445917bea2f62ccf3f2610acff92d8R36 of packing them to another props-like object, different form the original props. To me it introduces even more confusion: another data type to process in my brain, and think "is that format used expected by some other code as well?" "How those default values affect the flow, what arethe cases when that are different from the original props, so we don't do `getMinimumOrderHandlers( props )`?"

Reply:
I agree that 
```js
const { getA, getB, getC } = callGetABC( value, onChange );
// is shorter than 
const getA = callGetA( value, onChange );
const getB = callGetB( value, onChange );
const getC = callGetB( value, onChange );
```
but, personally, I don't think it's worth the complication of having:
>  a function, that given some args returns a function, that given another args returns a function, that is a handler, that given yet another args translates them and calls a callback given to the very first function.

https://github.com/woocommerce/google-listings-and-ads/compare/feature/refactor-minimum-order-modal-labelButton...feature/refactor-minimum-order-modal-handlers

Here is another proposal, to use just good, old, native `bind` to avoid the problem of a `getGetGetChangeHandler` and describing it.

--- 

Alternatively, we can just move the handlers' definitions back to `MinimumOrderCard` and we will make all those problems not applicable.
Then we could test the handlers' behavior via the component (which would test more of a component's behavior than implementation detail).

Reply:
> I think in most of the cases data come from React props, and I still don't see an actual benefit in [#1410 (files)](https://github.com/woocommerce/google-listings-and-ads/pull/1410/files#diff-19741a9dbe0895f99f4b04a6a88e21d3a8445917bea2f62ccf3f2610acff92d8R36) of packing them to another props-like object, different form the original props. [...]

Thanks, I think that sounds fair. I have made the changes accordingly in this commit: [`4c71bd9` (#1410)](https://github.com/woocommerce/google-listings-and-ads/pull/1410/commits/4c71bd931906fd376367f98c3b003b3c79627516).

I thought of a way to get around the hard problem of naming things 😂 , e.g. by using `onSubmit={ getMinimumOrderChangeHandler( value, onChange )( group ) }`. There is no "intermediary" function naming.

From the above, I also thought about changing the function signature to `getMinimumOrderChangeHandler( value, onChange, oldGroup, newGroup )`, similar to your code example. However, I don't quite like the way it looks. Maybe it's because I'm already used to the nested curried functions (like `() => () => {}`) where each "level" represent different things. With `getMinimumOrderChangeHandler( value, onChange )( oldGroup )( newGroup )`, the first level `value` and `onChange` relates to shipping rates that we received from props, the second level represents the rendered "oldGroup", and the third level represents the callback generated by the user action.

Let me know if you think it is okay. 🙏 

> Here is another proposal, to use just good, old, native `bind` to avoid the problem of a `getGetGetChangeHandler` and describing it.

💅 Personal preference: every time I see `.bind(null,...)`, I just feel like it is an eye sore and I don't like it. 😂  

---

> Alternatively, we can just move the handlers' definitions back to `MinimumOrderCard` and we will make all those problems not applicable.

I think that is a step backward. I like breaking things smaller (like what you have suggested for the handlers), and making the handlers outside the component allows us to test the handlers individually, which is good. 🙂 

> Then we could test the handlers' behavior via the component (which would test more of a component's behavior than implementation detail).

We can have handler testing and also component testing. I think they are / can be complementary to each other, and not necessarily one replacing another.

Reply:
>  I also thought about changing the function signature to `getMinimumOrderChangeHandler( value, onChange, oldGroup, newGroup )`, similar to your code example. However, I don't quite like the way it looks. Maybe it's because I'm already used to the nested curried functions (like `() => () => {}`) where each "level" represent different things. 

Wow, really? :) Personally, I don't like the way `getMinimumOrderChangeHandler( value, onChange )( oldGroup )( newGroup )` looks because I need to head scratch and read the docs to know:
- "What is the function that is the result of `getMinimumOrderChangeHandler( value, onChange )`?"
-  "What is that function meant for?"
-  "What is its signature?"
-  "What are its arguments?"
- "Is it meant to be used alone?"
- and the same story for `getMinimumOrderChangeHandler( value, onChange )( oldGroup )`

VSCode hint looks pretty complicated, and requires a handwritten description to make more sense:
![image](https://user-images.githubusercontent.com/17435/165982606-68f44380-8ac0-425b-ace4-311956cf7bc7.png)

While with `handleChange( value, onChange, oldGroup, newGroup )`:
- The IDE hint is straightforward
- There are no functions to name or ask for
- There is less code to implement those
- There are less function calls to be executed in run-time
- The stack trace is shorter and easier to reason, as there is one named function instead of a bunch of anonymous ones.

![image](https://user-images.githubusercontent.com/17435/165982453-08b42750-f6b2-4751-8562-27626e0a8752.png)

> With `getMinimumOrderChangeHandler( value, onChange )( oldGroup )( newGroup )`, the first level `value` and `onChange` relates to shipping rates that we received from props, the second level represents the rendered "oldGroup", and the third level represents the callback generated by the user action.

We have plenty of cases when the function argument values come from many sources, and I think it's not the reason to create more functions, just because someone who calls a function, may fetch arguments from different contexts.

-----

We never call `getMinimumOrderChangeHandler( value, onChange )` without the third argument, so I still don't see a reason to create another function and separate it there

-----

> Personal preference: every time I see `.bind(null,...)`, I just feel like it is an eye sore and I don't like it.

But this is a native function, part of the language for ages. Created specifically for that, so we don't have to "polyfill" it every time manually.

```
onChange={ handleChange.bind( null, value, onChange, groups[ 0 ] ) }
onChange={ ( newGroup ) => { handleChange( value, onChange, groups[ 0 ], newGroup ) }
onChange={ getMinimumOrderChangeHandler(
							value,
							onChange
						)( group[ 0 ] ) }
```

For me, the first is the easiest to read for me (plus the shortest and cheapest), and the latter is the most complicated to read, write, test, and execute.

Reply:
Thanka @tomalec , for your detailed explanations and bullet points! You shared some good perspectives there. I gave more thoughts about it and decided to change them and simplify them to `onChange={ getMinimumOrderChangeHandler( value, onChange, group ) }`, as commented in https://github.com/woocommerce/google-listings-and-ads/pull/1410#discussion_r862767458. 🙂 👍